Code review of existing code base

zhukovgreen opened the pull request:
Introduce a class with collections of utils to parse the date specified as a string

zhukovgreen committed:
Init

zhukovgreen committed:
Add basic test workflow

zhukovgreen committed:
Set UTC timezone

zhukovgreen committed:
Create README.md

zhukovgreen commented on src/test/scala/DateParserAppSuite.scala:
```diff
@@ -0,0 +1,32 @@
+      "should raise exceptions and log messages for bad inputs" in {
+        assertThrows[ParseException](date_parser.parse_date("yyyy-MM-dd", "aaag-12-12-12"))
+        assertThrows[IllegalArgumentException](date_parser.parse_date("ggg-ggg-ggg", "aaag-12-12-12"))
+        // TODO how to test log messages itself?
```
Any ideas?

yruslan replied:
We don't test log messages (e.g. messages produced by `log.info()` or similar), but we test messages generated by exceptions, like this:
```scala
    val ex = intercept[IllegalArgumentException] {
      doSomethingBad()
    }
    assert(ex.getMessage.contains("A part of the expected error message"))
```
We usually don't check the full message because the purpose here is that the expected exception was thrown on the expected behavior. Sometimes, if a message is complicated, we test message generation logic by comparing the full message of the exception.

zhukovgreen commented on src/test/scala/DateParserAppSuite.scala:
```diff
@@ -0,0 +1,32 @@
+  "DateParserAppSuite" - {
+    "date_parser" - {
+      "should parse good date with good format " in {
+        assert(date_parser.parse_date("yyyy-MM-dd", "2020-12-12").toString == "Sat Dec 12 00:00:00 UTC 2020")
```
Now tests fail locally (instead of UTC I have CET), how to compare the date better?

yruslan replied:
When you switch to Java Time you will have to provide a time zone to the formatter, so this shouldn't be an issue.

zhukovgreen reviewed (commented):
couple of notes

yruslan commented on src/main/scala/DateParserApp.scala:
```diff
@@ -0,0 +1,57 @@
+   * */
+  def parse_date(fmt: String, date: String): Date = {
+    try {
+      val date_formatter = new SimpleDateFormat(fmt)
```
Use `java.time.format.DateTimeFormatter` instead of `SimpleDateFormat`. SimpleDateFormat is an old and outdated class, in addition, it has concurrency issues. Java Time library is much better,

yruslan commented on src/main/scala/DateParserApp.scala:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @param log_level set logger level
+ * */
+class DateParserApp(log_level: LogLevel = LogLevel.INFO) extends LogSupport {
+
+  // Prepare logger
+  Logger.init
+  Logger.setDefaultLogLevel(log_level)
+  info(s"Logger set to ${Logger.getDefaultLogLevel} level")
```
1. We use `slf4j` api for logging. It is up to the runner to attach to the proper implementation.
2. Log levels should be set in config (log4j.properties, since Spark uses Log4J implementation), remove the log level from the constructor

The logger is almost always (with some very special exceptions to the rule) initialized as follows:
```scala
import org.slf4j.LoggerFactory

class SomeClass {
  private val log = LoggerFactory.getLogger(this.getClass)

  // The rest of the class implementation
}
```

zhukovgreen replied:
took this as nit

yruslan reviewed (commented):
This is a very good start. Despite my comments, the code is very good

yruslan commented on src/main/scala/DateParserApp.scala:
```diff
@@ -0,0 +1,57 @@
+    val maybe_result = POSSIBLE_DATA_FMTS.view.map(fmt => Try(parse_date(fmt, date))).find(_.isSuccess).map(_.get)
+    if (maybe_result.isEmpty) {
+      error(s"Can't parse the $date with the available formats: $POSSIBLE_DATA_FMTS")
+      throw new IllegalArgumentException
```
```suggestion
      throw new IllegalArgumentException(s"Can't parse the $date with the available formats: $POSSIBLE_DATA_FMTS")
```

yruslan commented on src/main/scala/DateParserApp.scala:
```diff
@@ -0,0 +1,57 @@
+   * @throws IllegalArgumentException if string can't be parsed for all known formats
+   * */
+  def guess_date(date: String): Date = {
+    val maybe_result = POSSIBLE_DATA_FMTS.view.map(fmt => Try(parse_date(fmt, date))).find(_.isSuccess).map(_.get)
```
```suggestion
    val maybe_result = POSSIBLE_DATA_FMTS.view.map(fmt => Try(parse_date(fmt, date))).find(_.isSuccess) match {
      case Some(parsedDate) => parsedDate
      case None => // Throw an exception
    }
```

zhukovgreen replied:
thx, very handy

yruslan commented on src/main/scala/DateParserApp.scala:
```diff
@@ -0,0 +1,57 @@
+    } catch {
+      case e: ParseException =>
+        error(s"Can't parse $date for $fmt")
+        throw e
```
```suggestion
        throw new IllegalArgumentException(s"Can't parse $date for $fmt", e)
```

yruslan commented on src/main/scala/DateParserApp.scala:
```diff
@@ -0,0 +1,57 @@
+  info(s"Logger set to ${Logger.getDefaultLogLevel} level")
+
+  // set some constants
+  val POSSIBLE_DATA_FMTS: List[String] = List("yyyy-MM-dd", "yyyy/MM/dd", "yyyy.MM.dd", "dd-MM-yyyy", "dd/MM/yyyy", "dd.MM.yyyy")
```
You can move this to the companion object

zhukovgreen committed:
Use java.time as date object +

Also improved check of exceptions in tests
Moved the constant to companion object

zhukovgreen committed:
Use pattern matching in guess_date

very clean

ZhukovGreen commented:
@yruslan pls glance one more time. It is definitely cleaner

zhukovgreen committed:
Add thorws to docstrings if parse_date

yruslan commented on src/main/scala/DateParserApp.scala:
```diff
@@ -0,0 +1,58 @@
+    } catch {
+      case _: DateTimeParseException =>
+        throw new DateTimeParseException(s"Can't parse $date for $fmt", s"$fmt", 1)
+      case _: IllegalArgumentException =>
+        throw new IllegalArgumentException(s"Wrong format: $fmt")
```
```suggestion
      case ex: IllegalArgumentException =>
        throw new IllegalArgumentException(s"Wrong format: $fmt", ex)
```

yruslan commented on src/main/scala/DateParserApp.scala:
```diff
@@ -0,0 +1,58 @@
+  def guess_date(date: String): LocalDate = {
+    DateParserApp.POSSIBLE_DATA_FMTS.view.map(fmt => Try(parse_date(fmt, date))).find(_.isSuccess).map(_.get) match {
+      case Some(result) => result
+      case None => throw new IllegalArgumentException(s"Can't parse the $date with the available formats: $DateParserApp.POSSIBLE_DATA_FMTS")
```
```suggestion
      case None => throw new IllegalArgumentException(s"Can't parse the $date with the available formats: ${DateParserApp.POSSIBLE_DATA_FMTS.mkString(", ")}")
```
